### PR TITLE
bugfix/S3C-4784 redis connection buildup

### PR DESCRIPTION
**Pull request by watsaqat**

### CODE REVIEW

File: tests/functional/client/testClientConnections.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    port: 16379,
+    host,
+});
+
+const sentinelSub = new Redis({
+    port: 16379,
+    host,
+});
+
+describe.only('Client connections', async function test() {
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: remove `.only`

File: tests/functional/client/testClientConnections.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        const MAX_RANGE_MS = (((1000 * 60) * 60) * 24) * 30; // One month.
+        const {
+            timeRange, type, resource, expected, action,
+        } = {
+            action: 'ListMetrics',
+            type: 'buckets',
+            resource: 'my-bucket',
+            timeRange: [
+                0,
+                ((MAX_RANGE_MS / 60) - (1000 * 60) * 15) - 1,
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: why `MAX_RANGE_MS / 60`?

**Review by tmacro — APPROVED** [review]

**Review by mrepop — APPROVED** [review]

File: tests/functional/client/testClientConnections.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            method: 'POST',
+            path: `/${type}?Action=${action}`,
+        };
+        const body = {
+            timeRange,
+            [type]: [resource],
+        };
+        ctx.requestsDuringFailover += 1;
+        makeUtapiGenericClientRequest(headers, body, (err, response) => {
+            if (err) return done(err);
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]:
```suggestion
            if (err) {
                return done(err);
            }
```

File: tests/functional/client/testClientConnections.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    afterEach(async () => {
+        const clients = await redis.client('list');
+        this.connectionsAfterFailover = clients.split('\n').length;
+        assert(this.connectionsAfterFailover <= this.connectionsBeforeFailover);
+    });
+
+    it('should not add connections after failover under load', done => {
+        sentinel.sentinel('failover', 'scality-s3', (err, res) => {
+            if (err) return done(err);
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]:
```suggestion
            if (err) {
                return done(err);
            }
```

File: tests/functional/client/testClientConnections.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            async.race([
+                () => setTimeout(() => this.loadgen.emit('finished'), 3000),
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: Is this to force stopping the ingestion after 3 seconds? Maybe a short explanation could help

File: tests/functional/client/testClientConnections.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                () => async.times(
+                    100,
+                    () => makeRequest(this, done),
+                    () => this.loadgen.emit('finished'),
+                ),
+            ]);
+
+            assert.strictEqual(res, 'OK');
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: Maybe putting this check after the `if (err)` block could be clearer

File: tests/functional/client/testClientConnections.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    host,
+});
+
+const sentinelSub = new Redis({
+    port: 16379,
+    host,
+});
+
+describe.only('Client connections', async function test() {
+    this.timeout(5000);
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: Is this timeout needed? I see you set it to 20 seconds in the only test that is part of this `describe` block

File: tests/functional/client/testClientConnections.js (hunk 2 above)
Comment (watsaqat) [reply tests/functional/client/testClientConnections.js]: I just copied this config from another test. It didn't matter too much what the requests were for this test as long as some load happened during a failover

File: tests/functional/client/testClientConnections.js [hunk 8]
```diff
@@ -105,7 +107,13 @@ describe.only('Client connections', async function test() {
...
     it('should not add connections after failover under load', done => {
         sentinel.sentinel('failover', 'scality-s3', (err, res) => {
-            if (err) return done(err);
+            if (err) {
+                return done(err);
+            }
+            assert.strictEqual(res, 'OK');
+
+            // Requests made with async.times will stay open when they occur
+            // during the failover window so and async.race is used to resolve
```
Comment (jonathan-gramain) [inline tests/functional/client/testClientConnections.js]: :+1:

**Review by jonathan-gramain — APPROVED** [review]

**Pull request closed without merging**